src/util/voidAreaDetection.ts

```diff
@@ -0,0 +1,106 @@
+ */
+const detectVoidAreaTap = (editable: HTMLElement, clientX: number, clientY: number): CaretPositionInfo | null => {
+  const doc = document as Document
+
+  // These APIs are not available in test environments (JSDOM)
+  // In that case, return null to let the browser handle it normally
+  if (!doc.caretRangeFromPoint && !doc.caretPositionFromPoint) {
+    return null
+  }
+
+  // Get the browser range for the tap position
+  let range: Range | null = null
+
+  if (doc.caretRangeFromPoint) {
+    range = doc.caretRangeFromPoint(clientX, clientY)
+  } else if (doc.caretPositionFromPoint) {
+    const pos = doc.caretPositionFromPoint(clientX, clientY)
+    if (pos?.offsetNode) {
+      range = document.createRange()
+      range.setStart(pos.offsetNode, pos.offset)
+      range.collapse(true)
+    }
+  }
+
+  // Ensure tap is within our editable element
+  if (!range || !editable.contains(range.startContainer)) {
+    return { node: null, nodeOffset: 0 }
+  }
+
+  const node = range.startContainer
+  const offset = range.startOffset
+  const nodeTextLength = node.textContent?.length || 0
+
+  /** Get the bounding rectangle for a character at the given offset. */
+  const getCharRect = (targetOffset: number): DOMRect | null => {
+    const charRange = document.createRange()
+    charRange.setStart(node, targetOffset)
+    charRange.setEnd(node, targetOffset + 1)
+    return charRange.getBoundingClientRect()
+  }
+
+  /** Check if tap is within a character's bounding box. */
+  const isInsideCharRect = (rect: DOMRect | null): boolean => {
+    if (!rect) return false
+    return clientX >= rect.left && clientX <= rect.right && clientY >= rect.top && clientY <= rect.bottom
+  }
+
+  /** Check if tap is vertically contained within the character's bounding box. */
+  const isVerticallyContained = (rect: DOMRect | null): boolean => {
+    if (!rect) return false
+    return clientY >= rect.top && clientY <= rect.bottom
+  }
+
+  // Check by tapping on a character at the current offset and the one before it.
+  const isClickOnCharacter = [offset, offset - 1]
+    .filter(o => o >= 0 && o < nodeTextLength)
+    .some(checkOffset => isInsideCharRect(getCharRect(checkOffset)))
+
+  // Allow taps horizontally beyond text if vertically aligned with text line
+  const isValidEdgeClick =
+    (offset === 0 || offset === nodeTextLength) &&
+    isVerticallyContained(getCharRect(offset === 0 ? 0 : nodeTextLength - 1))
+
+  // Valid tap on character, not a void area
+  if (isClickOnCharacter || isValidEdgeClick) return null
+
+  // Invalid tap (padding/void area), calculate the caret position
+  const textNodes = textNodeUtils.getTextNodes(editable)
+  if (textNodes.length === 0) return null
+
+  const nearest = textNodeUtils.findNearestTextNode(textNodes, clientY)
+  if (!nearest) return null
+
+  const nodeOffset = calculateHorizontalOffset(nearest.node, clientX, clientY)
+  return { node: nearest.node, nodeOffset }
+}
+
+export default detectVoidAreaTap
```

Comment:
In this project we follow the convention of:

1. Each module has a single, default export.
2. The file name matches the function name.

These conventions create a predictable relationship between function names, module names, and file names, making it easier to navigate the codebase and find where functions are located.